### DISCUSSION TITLE
Substituting placeholder variables in the text with values from a JSON

### DISCUSSION THREAD

**Discussion by michaelosthege** [executablebooks/meta#732] [Ideas]
Hi,

let's say I'm writing something on a computational method, or data analysis that is still under development.
And let's assume that the precise numbers (accuracy, probabilities, percentages, etc.) from my analysis are still changing, but the general conclusion remains the same.

(How) would it be possible to put **placeholder variables** in my text so I don't have to manually update the numbers each time they change in their 2nd decimal place?

For example:
```
The success rate with treatment A was {placeholder}`successRateA` % higher.
The blood concentration after 24 h was predicted at {placeholder}`bloodConc24h` mg/mL ± {placeholder}`bloodConc24herr`.
```
And a JSON file with the values, exported by the analysis pipeline:

```json
{
    "successRateA": 32.4,
    "bloodConc24h": 0.82,
    "bloodConc24herr": 0.03
}
```

I assume that this is already possible with custom LaTeX commands or so, but I wonder what would be the most elegant solution.

**Comment by schmoelder** [comment]
Does this help: https://jupyterbook.org/en/stable/content/executable/output-insert.html

**Comment by choldgraf** [comment]
Maybe some of the substitutions patterns would help? I think it might be tricky to define them programmatically, but this might be worth exploring: https://jupyterbook.org/en/stable/content/content-blocks.html?highlight=substitute#substitutions-and-variables-in-markdown

**Comment by michaelosthege** [comment]
Thanks, both of these links look good 👍
At this point in my writing I need to produce text, but I anticipate to adopt one of these substitution strategies at some point.

I'll come back to this thread when I can give a progress report.

__EDIT:__ Just linking this for later: https://myst-parser.readthedocs.io/en/latest/syntax/optional.html#substitutions-with-jinja2
...which also explains how the substitutions can be defined in `conf.py` so I should be easy to read them from a JSON or the like.
